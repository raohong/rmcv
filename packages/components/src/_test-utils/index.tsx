import createCache from '@emotion/cache';
import { CacheProvider } from '@emotion/react';
import { RenderOptions, render } from '@testing-library/react';
import React, { ReactElement } from 'react';

export { default as sleep } from './sleep';

const emotionCache = createCache({ stylisPlugins: [], key: 'test' });
const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return <CacheProvider value={emotionCache}>{children}</CacheProvider>;
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
