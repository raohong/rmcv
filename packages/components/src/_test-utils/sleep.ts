const sleep = (timeout = 0): Promise<void> =>
  new Promise((r) =>
    setTimeout(() => {
      r();
    }, timeout),
  );

export default sleep;
