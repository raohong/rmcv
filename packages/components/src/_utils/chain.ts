const chain = <T extends (...args: any[]) => void>(
  ...fns: (T | undefined)[]
): ((...args: Parameters<T>) => void) => {
  return (...args) => {
    fns.forEach((item) => {
      if (item) {
        item(...args);
      }
    });
  };
};

export default chain;
