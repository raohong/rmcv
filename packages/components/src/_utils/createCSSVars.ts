import kebabCase from 'lodash/kebabCase';

const createCSSVars = (
  prefix: string,
  style: Record<string, string | undefined | null>,
) => {
  const result = Object.entries(style).reduce((map, [key, value]) => {
    if (!value) {
      return map;
    }
    // @ts-ignore
    // eslint-disable-next-line no-param-reassign
    map[`--${prefix}-${kebabCase(key)}`] = value;

    return map;
  }, {} as React.CSSProperties);

  return result;
};

export default createCSSVars;
