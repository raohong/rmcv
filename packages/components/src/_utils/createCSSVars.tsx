import React from 'react';

export const createCSSVars = (vars: Record<string, number | string>) =>
  Object.entries(vars).reduce((result, [name, value]) => {
    // @ts-ignore
    result[`--${name}`] = value;
    return result;
  }, {} as React.CSSProperties);
