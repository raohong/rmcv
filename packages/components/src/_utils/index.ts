export { default as getPrefixCls } from './getPrefixCls';

export { default as isBrowser } from './isBrowser';

export { default as renderPortal } from './renderPortal';

export { default as uuid } from './uuid';

export { default as createCSSVars } from './createCSSVars';

export { default as chain } from './chain';

export { default as toArray } from './toArray';
