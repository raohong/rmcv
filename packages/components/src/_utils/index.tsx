export * from './createOverridableComponent';
export * from './decay';
export * from './flatReactNode';
export * from './getPrefixCls';
export * from './htmlAttrs';
export * from './reactRender';
export * from './useXId';
