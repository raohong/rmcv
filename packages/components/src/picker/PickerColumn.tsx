import { useSpring } from '@react-spring/web';
import { useUpdateEffect } from '@rmc-vant/hooks';
import { clamp, isNumber } from '@rmc-vant/utils';
import { useDrag } from '@use-gesture/react';
import { useRef } from 'react';
import { getDataOrAriaProps } from '../_utils';
import type { PickerColumnProps, PickerValue } from './interface';
import { PickerColumnRoot, PickerColumnWrapper, PickerOption } from './styles';

const PickerColumn = <V extends PickerValue>({
  onChange,
  selectedIndex,
  optionHeight,
  className,
  options,
  immediateChange,
  columnIndex,
  totalHeight,
  slotClassNames,
  componentState,
  ...rest
}: PickerColumnProps<V>) => {
  const ref = useRef<HTMLDivElement>(null);
  const [{ y }, ctrl] = useSpring(() => ({
    y: isNumber(selectedIndex) ? selectedIndex * -optionHeight : 0,
  }));
  const lastVelocity = useRef(0);

  const { length } = options;
  const bounds = [-optionHeight * (length - 1), 0];

  const handleChange = (targetIndex: number, velocity: number = 0) => {
    const value = options[targetIndex]?.value as V;

    if (immediateChange) {
      onChange?.(columnIndex, value);
      lastVelocity.current = velocity;
    }
    else {
      ctrl.start({
        y: targetIndex * -optionHeight,
        config: {
          velocity,
        },
        onRest({ finished }) {
          if (!immediateChange && finished) {
            onChange?.(columnIndex, value);
          }
        },
      });
    }
  };

  useDrag(
    ({ offset: [, cy], last, tap, velocity: [, vy], direction: [, dy] }) => {
      if (tap || selectedIndex === undefined) {
        return;
      }

      if (!last) {
        ctrl.set({ y: cy });
        return;
      }

      const power = 0.7;
      const timeConst = 350;
      let index = selectedIndex;

      let target = cy;
      if (cy > bounds[0] && cy < bounds[1]) {
        target = timeConst * power * vy * dy + cy;
      }

      target = clamp(target, (length - 1) * -optionHeight, 0);
      index = Math.floor(Math.abs(target) / optionHeight);

      if (options[index]?.disabled) {
        index = selectedIndex;
      }

      if (index === selectedIndex) {
        ctrl.start({
          y: index * -optionHeight,
          config: {
            velocity: vy,
          },
        });
      }
      else {
        handleChange(index, vy);
      }
    },
    {
      target: ref,
      from: () => [0, y.get()],
      axis: 'y',
      filterTaps: true,
      eventOptions: {
        passive: false,
      },
      rubberband: 0.5,
      bounds: {
        top: bounds[0],
        bottom: bounds[1],
      },
      pointer: {
        touch: true,
      },
      preventDefault: true,
    },
  );

  useUpdateEffect(() => {
    if (
      optionHeight > 0
      && isNumber(selectedIndex)
      && y.get() !== -optionHeight * selectedIndex
    ) {
      ctrl.start({
        y: -optionHeight * selectedIndex,
        config: {
          velocity: lastVelocity.current,
        },
      });
      lastVelocity.current = 0;
    }
  }, [optionHeight, selectedIndex, ctrl, y]);

  return (
    <PickerColumnRoot
      componentState={componentState}
      ref={ref}
      style={{ height: totalHeight }}
      {...getDataOrAriaProps(rest)}
    >
      <PickerColumnWrapper
        style={{
          y,
          top: `calc(50% - ${optionHeight / 2}px)`,
        }}
      >
        {options.map((item, index) => (
          <PickerOption
            role='button'
            componentState={{
              ...componentState,
              disabled: !!item.disabled,
            }}
            tabIndex={0}
            aria-disabled={item.disabled || undefined}
            data-selected={index === selectedIndex || undefined}
            onClick={
              item.disabled
                ? undefined
                : () => {
                    handleChange(index);
                  }
            }
            style={{ height: optionHeight }}
            key={item.value}
            className={slotClassNames.option}
          >
            {item.label ?? item.value}
          </PickerOption>
        ))}
      </PickerColumnWrapper>
    </PickerColumnRoot>
  );
};

export default PickerColumn;
