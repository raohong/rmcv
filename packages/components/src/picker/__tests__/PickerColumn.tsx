import React from 'react';
import { screen, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import PickerColumn from '../PickerColumn';
import { getPrefixCls } from '../../_utils';
import { sleep } from '../../_test-utils';

const optionHeight = 44;
const num = 6;
const options = ['A', 'B', 'C'].map((item) => ({ value: item }));
const totalHeight = num * optionHeight;

const testId = 'picker-column';

test('render correctly', () => {
  const tree = render(
    <PickerColumn
      columnIndex={0}
      totalHeight={totalHeight}
      optionHeight={44}
      options={options}
    />,
  );

  expect(tree.asFragment()).toMatchSnapshot();
});

test('trigger onChange by click', async () => {
  const onChange = jest.fn();
  const com = render(
    <PickerColumn
      columnIndex={0}
      totalHeight={totalHeight}
      optionHeight={44}
      options={options}
      selectedIndex={0}
      onChange={onChange}
      data-testid={testId}
    />,
  );

  const option = screen
    .getByTestId(testId)
    .querySelector(`.${getPrefixCls('picker-option')}:nth-child(2)`);

  userEvent.click(option!);

  await sleep(1000);

  expect(onChange).toBeCalled();

  com.rerender(
    <PickerColumn
      columnIndex={0}
      totalHeight={totalHeight}
      optionHeight={44}
      options={options}
      selectedIndex={0}
      onChange={onChange}
      data-testid={testId}
      immediateChange
    />,
  );

  userEvent.click(option!);

  expect(onChange).toBeCalled();
});
