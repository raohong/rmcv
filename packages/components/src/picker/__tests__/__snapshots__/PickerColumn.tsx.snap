// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  <div
    class="rmcv-picker-column"
    style="height: 264px;"
  >
    <div
      class="rmcv-picker-column-slider"
    >
      <ul
        class="rmcv-picker-column-slider-content"
        style="top: calc(50% - 22px); transform: none;"
      >
        <li
          class="rmcv-picker-option"
          role="button"
          style="height: 44px;"
          tabindex="0"
        >
          A
        </li>
        <li
          class="rmcv-picker-option"
          role="button"
          style="height: 44px;"
          tabindex="0"
        >
          B
        </li>
        <li
          class="rmcv-picker-option"
          role="button"
          style="height: 44px;"
          tabindex="0"
        >
          C
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
